### ISSUE TITLE
Automatic cluster bootstrap

### ISSUE THREAD

**Issue by rtsisyk** [tarantool/tarantool#1062]
The proposal allow to use the same configuration file on all cluster nodes for async/bsync replication. Described approach eliminates needs in a special bootstrap procedure for the "master" node. The only bootstrap logic is changed. 

## Bootstrap Workflow

These logic applies only when **recovery is empty** (i.e. it doesn't have snapshots and xlogs) and box.cfg {} has **at least one replica**. In other cases nothing is changed.

1. Generate server UUID

2. Start IPROTO listener and return UUID in greeting.
All requests except **AUTH** should raise an exception (e.g ER_BOOTSTRAP). iproto.cc code must be functional without fully initialized box. 

3.  Create `struct applier` for all replicas in `box.cfg {}` and try to connect. You can do this step in parallel or synchronously - this doesn't matter for the spec.

4. If IPROTO greeting UUID **is not equal** to local UUID:
   1. Send AUTH request as usual (if needed)
   2. Send JOIN request as usual
   3. On `ER_BOOTSTRAP` on any of these steps sleep and try again later. Please don't print extra error messages for ER_BOOTSTRAP.

5. If IPROTO greeting UUID **is equal** to local UUID:
   1. **[bsync only]** Save user and password from URI to a global variables e.g. `bootstrap_user` and `bootstrap_password`. **Please distinguish between case then "user and password is not set" and "user and password is not known yet".**
   2. Set flag in `struct applier` that connection is local and then disconnect
   3. Create an initial snapshot if this **replica is first in the set** (sort the set by URI). In case of bsync we will do leader election here. **Switch IPROTO to regular mode.** Process other requests as usual.
  
## AUTH request handling 

1. **[bsync only]** if global variable `bootstrap_user` and `bootstrap_password` is set, check credentials and authorize user. **Please distinguish between the case when "user and password are not set" and "user and password are not known yet".**
2. Otherwise return `ER_BOOTSTRAP `- at this step master can't check client credentials. Please note that `ER_PASSWORD_MISMATCH` is invalid here, because we actually don't know whether credentials are valid or not.

## Other request handling

1. Always return `ER_BOOTSTRAP` until snapshot is generated or received from the master.

Please note that the only first replica in the replica set can be explicitly authorized by checking UUID in greeting. `bootstrap_user` and `bootstap_password` is only needed for bsync. IPROTO in regular mode must not allow `AUTH` with `bootstrap_user` and `bootstap_password` credentials. 

This feature can be implemented separately and pushed to 1.6.x without breaking compatibility. Client should handle server errors in any case, so a new error code sent during bootstrap should not be a problem.

**Comment by rtsisyk** [comment]
Compatibility with #999: Don't bind to the port before downloading all snapshots from the master:
Just return `ER_BOOTSTRAP` for cases then replica has fallen a lot behind the master.

**Comment by rtsisyk** [comment]
Potential pitfall: replicas should iterate over connections during JOIN.
Current implementation was n fibers that tries to do JOIN in parallel.

**Comment by rtsisyk** [comment]
Epic fail - UUID doesn't fit to greeting.

```
>>> 63 - len('Tarantool 1.6.7-194-g8f1b153 (Lua console)')
21
```

**Comment by rtsisyk** [comment]
My branch proably should work after rebasing on the latest 1.7.x.
I need time to try.

**Comment by rtsisyk** [comment]
Blocked by #1114 

Current implementation has broken box.once() because replica doesn't wait for xlogs in JOIN:

```
box.cfg({
    listen = replicauri(INSTANCE_ID);
    log_level = 7;
    replication_source = {
        replicauri(1);
        replicauri(2);
        replicauri(3);
    };
})
box.once("bootstrap", function()
    box.schema.user.grant('guest','read,write,execute','universe')
    box.schema.space.create('test')
    box.space.test:create_index('primary')
end)
```

**Comment by rtsisyk** [comment]
Fixed by 1.7.0-795-g70e6e55
